launcher/src/code-workspace.ts

```diff
@@ -30,54 +44,103 @@ export class CodeWorkspace {
       return;
     }
 
-    // skip if env.VSCODE_DEFAULT_WORKSPACE is defined and points to a real file
-    if (env.VSCODE_DEFAULT_WORKSPACE) {
-      if (
-        (await fs.pathExists(env.VSCODE_DEFAULT_WORKSPACE)) &&
-        (await fs.isFile(env.VSCODE_DEFAULT_WORKSPACE))
-      ) {
-        console.log(
-          "  > env.VSCODE_DEFAULT_WORKSPACE is defined, skip this step"
-        );
-        return;
-      } else {
-        console.log(
-          "  > env.VSCODE_DEFAULT_WORKSPACE must point on a workspace file"
-        );
-      }
-    }
+    // Following flow is common for default workspce location and
+    // for custom workspace file defined in env.VSCODE_DEFAULT_WORKSPACE
+    const alternativePath = env.VSCODE_DEFAULT_WORKSPACE;
 
     try {
-      // skip if workspace file is already exist
-      if (await fs.pathExists(workspaceFilePath())) {
-        console.log(
-          `  > Workspace file ${workspaceFilePath()} is already exist`
-        );
-        return;
-      }
+      // get workspace file
+      const workspace = await this.readWorkspaceFile(alternativePath);
 
-      // take all the projects from flattened devworkspace file
-      const projects = await new FlattenedDevfile().getProjects();
-      const folders = projects.map((p) => {
-        return {
-          name: p.name,
-          path: `${env.PROJECTS_ROOT}/${p.name}`,
-        };
-      });
-
-      const workspace = {
-        folders,
-      };
-
-      // write .code-workspace file
-      const json = JSON.stringify(workspace, null, "\t");
-      await fs.writeFile(workspaceFilePath(), json);
+      const devfile = await new FlattenedDevfile().getDevfile();
 
-      console.log(`  > writing ${workspaceFilePath()}..`);
+      // synchronize projects
+      this.synchronizeProjects(workspace, devfile.projects);
+
+      // synchronize dependent projects
+      this.synchronizeProjects(workspace, devfile.dependentProjects);
+
+      // synchronize starter projects
+      this.synchronizeProjects(workspace, devfile.starterProjects);
+
+      // write workspace file
+      await this.writeWorkspaceFile(workspace, alternativePath);
     } catch (err) {
       console.error(
         `${err.message} Unable to generate che.code-workspace file`
       );
     }
   }
+
+  async isExist(workspaceFilePath: string | undefined): Promise<boolean> {
+    if (
+      workspaceFilePath &&
+      (await fs.pathExists(workspaceFilePath)) &&
+      (await fs.isFile(workspaceFilePath))
+    ) {
+      return true;
+    }
+
+    return false;
+  }
+
+  // should validate the content of read file
+  async readWorkspaceFile(alternativePath?: string): Promise<Workspace> {
```

Comment:
Can we simplify it a little?

the current state is: 
- the method gets `alternativePath` as a parameter => tries to read a file
- if the path is absent => tries to read another path like `${env.PROJECTS_ROOT}/.code-workspace`
- but actually `alternativePath` comes as `env.VSCODE_DEFAULT_WORKSPACE`

From my point of view it's better:
- to have a function that returns `workspaceFilePath`
- the function should return `alternativePath` if it exists, so `env.VSCODE_DEFAULT_WORKSPACE`
- otherwise the function returns `${env.PROJECTS_ROOT}/.code-workspace`

wdyt?

Reply:
btw: `alternativePath` - why it's `alternative` - should it be `defaultWorkspacePath` as it comes as `env.VSCODE_DEFAULT_WORKSPACE`?

Reply:
Default workspace file path is `${env.PROJECTS_ROOT}/.code-workspace`, it is used if nothing else is defined.
Despite the variable named as `VSCODE_DEFAULT_WORKSPACE`, for us it's the alternative location.
I simplified the method a bit.